### ISSUE TITLE
BUG: Idefix is not compiling when setting the Idefix_MPI option in a local CMakeLists.txt

### ISSUE THREAD

**Issue by jeankpf** [idefix-code/idefix#176]
### Describe the issue:

Tested with v2.0 commit f02fb501073f088ce7933bc075d066dfe778cf5d (latest one, branch downloaded today)
Basically this is because, when doing so, the Idefix_MPI option in $IDEFIX_DIR/src/CMakeLists.txt (where MPI files are included) (and a fortiori in $IDEFIX_DIR/CMakeLists.txt) is still set to False when the subdirectory $IDEFIX_DIR/src is called from $IDEFIX_DIR/CMakeLists.txt on line 61 with the instruction 'add_subdirectory(src build)' while the Idefix_MPI option is only set to True lines 71-74.
I don't know if this behavior is actually intended or not, but if not, a simple fix consists in placing the instruction currently on line 61 after the block of instructions currently on lines 71-74.

### Error message:

```shell
[100%] Linking CXX executable idefix
/usr/bin/ld: CMakeFiles/idefix.dir/src/dataBlock/dataBlock.cpp.o: in function `DataBlock::DataBlock(Grid&, Input&)':
dataBlock.cpp:(.text+0x6b77): undefined reference to `Mpi::~Mpi()'
/usr/bin/ld: dataBlock.cpp:(.text+0x75de): undefined reference to `Mpi::~Mpi()'
/usr/bin/ld: CMakeFiles/idefix.dir/src/dataBlock/dataBlock.cpp.o: in function `std::unique_ptr<Fargo, std::default_delete<Fargo> >::~unique_ptr()':
dataBlock.cpp:(.text._ZNSt10unique_ptrI5FargoSt14default_deleteIS0_EED2Ev[_ZNSt10unique_ptrI5FargoSt14default_deleteIS0_EED5Ev]+0x6a): undefined reference to `Mpi::~Mpi()'
/usr/bin/ld: CMakeFiles/idefix.dir/src/dataBlock/dataBlock.cpp.o: in function `std::unique_ptr<Gravity, std::default_delete<Gravity> >::~unique_ptr()':
dataBlock.cpp:(.text._ZNSt10unique_ptrI7GravitySt14default_deleteIS0_EED2Ev[_ZNSt10unique_ptrI7GravitySt14default_deleteIS0_EED5Ev]+0x23c): undefined reference to `Mpi::~Mpi()'
/usr/bin/ld: CMakeFiles/idefix.dir/src/dataBlock/dataBlock.cpp.o: in function `std::unique_ptr<Boundary<DefaultPhysics>, std::default_delete<Boundary<DefaultPhysics> > >::~unique_ptr()':
dataBlock.cpp:(.text._ZNSt10unique_ptrI8BoundaryI14DefaultPhysicsESt14default_deleteIS2_EED2Ev[_ZNSt10unique_ptrI8BoundaryI14DefaultPhysicsESt14default_deleteIS2_EED5Ev]+0x145): undefined reference to `Mpi::~Mpi()'
/usr/bin/ld: CMakeFiles/idefix.dir/src/dataBlock/dataBlock.cpp.o:dataBlock.cpp:(.text._ZNSt10unique_ptrI10RKLegendreI14DefaultPhysicsESt14default_deleteIS2_EED2Ev[_ZNSt10unique_ptrI10RKLegendreI14DefaultPhysicsESt14default_deleteIS2_EED5Ev]+0x19): more undefined references to `Mpi::~Mpi()' follow
/usr/bin/ld: CMakeFiles/idefix.dir/src/dataBlock/dataBlock.cpp.o: in function `Boundary<DustPhysics>::SetBoundaries(double)':
dataBlock.cpp:(.text._ZN8BoundaryI11DustPhysicsE13SetBoundariesEd[_ZN8BoundaryI11DustPhysicsE13SetBoundariesEd]+0x185): undefined reference to `Mpi::ExchangeX1(Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>, Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>)'
/usr/bin/ld: dataBlock.cpp:(.text._ZN8BoundaryI11DustPhysicsE13SetBoundariesEd[_ZN8BoundaryI11DustPhysicsE13SetBoundariesEd]+0x23d): undefined reference to `Mpi::ExchangeX3(Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>, Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>)'
/usr/bin/ld: dataBlock.cpp:(.text._ZN8BoundaryI11DustPhysicsE13SetBoundariesEd[_ZN8BoundaryI11DustPhysicsE13SetBoundariesEd]+0x307): undefined reference to `Mpi::ExchangeX2(Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>, Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>)'
/usr/bin/ld: CMakeFiles/idefix.dir/src/dataBlock/dataBlock.cpp.o: in function `Boundary<DefaultPhysics>::SetBoundaries(double)':
dataBlock.cpp:(.text._ZN8BoundaryI14DefaultPhysicsE13SetBoundariesEd[_ZN8BoundaryI14DefaultPhysicsE13SetBoundariesEd]+0x185): undefined reference to `Mpi::ExchangeX1(Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>, Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>)'
/usr/bin/ld: dataBlock.cpp:(.text._ZN8BoundaryI14DefaultPhysicsE13SetBoundariesEd[_ZN8BoundaryI14DefaultPhysicsE13SetBoundariesEd]+0x23d): undefined reference to `Mpi::ExchangeX3(Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>, Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>)'
/usr/bin/ld: dataBlock.cpp:(.text._ZN8BoundaryI14DefaultPhysicsE13SetBoundariesEd[_ZN8BoundaryI14DefaultPhysicsE13SetBoundariesEd]+0x307): undefined reference to `Mpi::ExchangeX2(Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>, Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>)'
/usr/bin/ld: CMakeFiles/idefix.dir/src/dataBlock/dataBlock.cpp.o: in function `RKLegendre<DefaultPhysics>::~RKLegendre()':
dataBlock.cpp:(.text._ZN10RKLegendreI14DefaultPhysicsED2Ev[_ZN10RKLegendreI14DefaultPhysicsED5Ev]+0x10): undefined reference to `Mpi::~Mpi()'
/usr/bin/ld: CMakeFiles/idefix.dir/src/dataBlock/dataBlock.cpp.o: in function `RKLegendre<DustPhysics>::~RKLegendre()':
dataBlock.cpp:(.text._ZN10RKLegendreI11DustPhysicsED2Ev[_ZN10RKLegendreI11DustPhysicsED5Ev]+0x10): undefined reference to `Mpi::~Mpi()'
/usr/bin/ld: CMakeFiles/idefix.dir/src/dataBlock/dataBlock.cpp.o: in function `Fluid<DefaultPhysics>::~Fluid()':
dataBlock.cpp:(.text._ZN5FluidI14DefaultPhysicsED2Ev[_ZN5FluidI14DefaultPhysicsED5Ev]+0x42d): undefined reference to `Mpi::~Mpi()'
/usr/bin/ld: dataBlock.cpp:(.text._ZN5FluidI14DefaultPhysicsED2Ev[_ZN5FluidI14DefaultPhysicsED5Ev]+0x611): undefined reference to `Mpi::~Mpi()'
/usr/bin/ld: CMakeFiles/idefix.dir/src/dataBlock/dataBlock.cpp.o: in function `Fluid<DustPhysics>::~Fluid()':
dataBlock.cpp:(.text._ZN5FluidI11DustPhysicsED2Ev[_ZN5FluidI11DustPhysicsED5Ev]+0x42d): undefined reference to `Mpi::~Mpi()'
/usr/bin/ld: CMakeFiles/idefix.dir/src/dataBlock/dataBlock.cpp.o:dataBlock.cpp:(.text._ZN5FluidI11DustPhysicsED2Ev[_ZN5FluidI11DustPhysicsED5Ev]+0x611): more undefined references to `Mpi::~Mpi()' follow
/usr/bin/ld: CMakeFiles/idefix.dir/src/dataBlock/dataBlock.cpp.o: in function `Boundary<DustPhysics>::Boundary(Fluid<DustPhysics>*)':
dataBlock.cpp:(.text._ZN8BoundaryI11DustPhysicsEC2EP5FluidIS0_E[_ZN8BoundaryI11DustPhysicsEC5EP5FluidIS0_E]+0x5af): undefined reference to `Mpi::Init(Grid*, std::vector<int, std::allocator<int> >, int*, int*, bool)'
/usr/bin/ld: dataBlock.cpp:(.text._ZN8BoundaryI11DustPhysicsEC2EP5FluidIS0_E[_ZN8BoundaryI11DustPhysicsEC5EP5FluidIS0_E]+0xba3): undefined reference to `Mpi::~Mpi()'
/usr/bin/ld: CMakeFiles/idefix.dir/src/dataBlock/dataBlock.cpp.o: in function `RKLegendre<DefaultPhysics>::RKLegendre(Input&, Fluid<DefaultPhysics>*)':
dataBlock.cpp:(.text._ZN10RKLegendreI14DefaultPhysicsEC2ER5InputP5FluidIS0_E[_ZN10RKLegendreI14DefaultPhysicsEC5ER5InputP5FluidIS0_E]+0x7c9): undefined reference to `Mpi::Init(Grid*, std::vector<int, std::allocator<int> >, int*, int*, bool)'
/usr/bin/ld: dataBlock.cpp:(.text._ZN10RKLegendreI14DefaultPhysicsEC2ER5InputP5FluidIS0_E[_ZN10RKLegendreI14DefaultPhysicsEC5ER5InputP5FluidIS0_E]+0x1454): undefined reference to `Mpi::~Mpi()'
/usr/bin/ld: CMakeFiles/idefix.dir/src/dataBlock/dataBlock.cpp.o: in function `RKLegendre<DustPhysics>::RKLegendre(Input&, Fluid<DustPhysics>*)':
dataBlock.cpp:(.text._ZN10RKLegendreI11DustPhysicsEC2ER5InputP5FluidIS0_E[_ZN10RKLegendreI11DustPhysicsEC5ER5InputP5FluidIS0_E]+0x7c9): undefined reference to `Mpi::Init(Grid*, std::vector<int, std::allocator<int> >, int*, int*, bool)'
/usr/bin/ld: dataBlock.cpp:(.text._ZN10RKLegendreI11DustPhysicsEC2ER5InputP5FluidIS0_E[_ZN10RKLegendreI11DustPhysicsEC5ER5InputP5FluidIS0_E]+0x1454): undefined reference to `Mpi::~Mpi()'
/usr/bin/ld: CMakeFiles/idefix.dir/src/dataBlock/dataBlock.cpp.o: in function `Boundary<DefaultPhysics>::Boundary(Fluid<DefaultPhysics>*)':
dataBlock.cpp:(.text._ZN8BoundaryI14DefaultPhysicsEC2EP5FluidIS0_E[_ZN8BoundaryI14DefaultPhysicsEC5EP5FluidIS0_E]+0x5af): undefined reference to `Mpi::Init(Grid*, std::vector<int, std::allocator<int> >, int*, int*, bool)'
/usr/bin/ld: dataBlock.cpp:(.text._ZN8BoundaryI14DefaultPhysicsEC2EP5FluidIS0_E[_ZN8BoundaryI14DefaultPhysicsEC5EP5FluidIS0_E]+0xba3): undefined reference to `Mpi::~Mpi()'
/usr/bin/ld: CMakeFiles/idefix.dir/src/dataBlock/dataBlock.cpp.o: in function `Fluid<DefaultPhysics>::Fluid(Grid&, Input&, DataBlock*, int)':
dataBlock.cpp:(.text._ZN5FluidI14DefaultPhysicsEC2ER4GridR5InputP9DataBlocki[_ZN5FluidI14DefaultPhysicsEC5ER4GridR5InputP9DataBlocki]+0x1935): undefined reference to `Mpi::~Mpi()'
/usr/bin/ld: CMakeFiles/idefix.dir/src/dataBlock/dataBlock.cpp.o: in function `Fluid<DustPhysics>::Fluid(Grid&, Input&, DataBlock*, int)':
dataBlock.cpp:(.text._ZN5FluidI11DustPhysicsEC2ER4GridR5InputP9DataBlocki[_ZN5FluidI11DustPhysicsEC5ER4GridR5InputP9DataBlocki]+0x18e5): undefined reference to `Mpi::~Mpi()'
/usr/bin/ld: CMakeFiles/idefix.dir/src/dataBlock/evolveStage.cpp.o: in function `RKLegendre<DefaultPhysics>::SetBoundaries(double)':
evolveStage.cpp:(.text._ZN10RKLegendreI14DefaultPhysicsE13SetBoundariesEd[_ZN10RKLegendreI14DefaultPhysicsE13SetBoundariesEd]+0x194): undefined reference to `Mpi::ExchangeX1(Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>, Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>)'
/usr/bin/ld: evolveStage.cpp:(.text._ZN10RKLegendreI14DefaultPhysicsE13SetBoundariesEd[_ZN10RKLegendreI14DefaultPhysicsE13SetBoundariesEd]+0x260): undefined reference to `Mpi::ExchangeX3(Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>, Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>)'
/usr/bin/ld: evolveStage.cpp:(.text._ZN10RKLegendreI14DefaultPhysicsE13SetBoundariesEd[_ZN10RKLegendreI14DefaultPhysicsE13SetBoundariesEd]+0x341): undefined reference to `Mpi::ExchangeX2(Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>, Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>)'
/usr/bin/ld: CMakeFiles/idefix.dir/src/dataBlock/fargo.cpp.o: in function `Fargo::Fargo(Input&, int, DataBlock*)':
fargo.cpp:(.text+0x4065): undefined reference to `Mpi::Init(Grid*, std::vector<int, std::allocator<int> >, int*, int*, bool)'
/usr/bin/ld: CMakeFiles/idefix.dir/src/dataBlock/fargo.cpp.o: in function `Fargo::Fargo(Input&, int, DataBlock*) [clone .cold]':
fargo.cpp:(.text.unlikely+0x5c3): undefined reference to `Mpi::~Mpi()'
/usr/bin/ld: CMakeFiles/idefix.dir/src/dataBlock/fargo.cpp.o: in function `void Fargo::StoreToScratch<DefaultPhysics>(Fluid<DefaultPhysics>*)':
fargo.cpp:(.text._ZN5Fargo14StoreToScratchI14DefaultPhysicsEEvP5FluidIT_E[_ZN5Fargo14StoreToScratchI14DefaultPhysicsEEvP5FluidIT_E]+0x53c): undefined reference to `Mpi::ExchangeX2(Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>, Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>)'
/usr/bin/ld: CMakeFiles/idefix.dir/src/dataBlock/fargo.cpp.o: in function `void Fargo::StoreToScratch<DustPhysics>(Fluid<DustPhysics>*)':
fargo.cpp:(.text._ZN5Fargo14StoreToScratchI11DustPhysicsEEvP5FluidIT_E[_ZN5Fargo14StoreToScratchI11DustPhysicsEEvP5FluidIT_E]+0x53c): undefined reference to `Mpi::ExchangeX2(Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>, Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>)'
/usr/bin/ld: CMakeFiles/idefix.dir/src/gravity/gravity.cpp.o: in function `SelfGravity::~SelfGravity()':
gravity.cpp:(.text._ZN11SelfGravityD2Ev[_ZN11SelfGravityD5Ev]+0x21d): undefined reference to `Mpi::~Mpi()'
/usr/bin/ld: gravity.cpp:(.text._ZN11SelfGravityD2Ev[_ZN11SelfGravityD5Ev]+0x2ae): undefined reference to `Mpi::~Mpi()'
/usr/bin/ld: CMakeFiles/idefix.dir/src/gravity/selfGravity.cpp.o: in function `SelfGravity::SetBoundaries(Kokkos::View<double***, Kokkos::LayoutRight, Kokkos::Serial>&)':
selfGravity.cpp:(.text+0xafc1): undefined reference to `Mpi::ExchangeX1(Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>, Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>)'
/usr/bin/ld: selfGravity.cpp:(.text+0xb06b): undefined reference to `Mpi::ExchangeX3(Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>, Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>)'
/usr/bin/ld: selfGravity.cpp:(.text+0xb12b): undefined reference to `Mpi::ExchangeX2(Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>, Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Serial>)'
/usr/bin/ld: CMakeFiles/idefix.dir/src/gravity/selfGravity.cpp.o: in function `SelfGravity::Init(Input&, DataBlock*)':
selfGravity.cpp:(.text+0x11f37): undefined reference to `Mpi::Init(Grid*, std::vector<int, std::allocator<int> >, int*, int*, bool)'
/usr/bin/ld: CMakeFiles/idefix.dir/src/global.cpp.o: in function `idfx::initialize()':
global.cpp:(.text+0x50): undefined reference to `Mpi::CheckConfig()'
/usr/bin/ld: CMakeFiles/idefix.dir/src/main.cpp.o: in function `DataBlock::~DataBlock()':
main.cpp:(.text._ZN9DataBlockD2Ev[_ZN9DataBlockD5Ev]+0x24c): undefined reference to `Mpi::~Mpi()'
/usr/bin/ld: main.cpp:(.text._ZN9DataBlockD2Ev[_ZN9DataBlockD5Ev]+0x2e4): undefined reference to `Mpi::~Mpi()'
/usr/bin/ld: CMakeFiles/idefix.dir/src/main.cpp.o: in function `main.cold':
main.cpp:(.text.unlikely+0x11f): undefined reference to `Mpi::CheckSync(double)'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/idefix.dir/build.make:597: idefix] Error 1
make[1]: *** [CMakeFiles/Makefile2:571: CMakeFiles/idefix.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

### runtime information:

commit f02fb501073f088ce7933bc075d066dfe778cf5d
error reproduced on my local computer (Dell) and on Olympe (Calmip mesocentre )

**Comment by jeankpf** [comment]
Update following @neutrinoceros crash course's on how to use GitHub properly (thanks again!). Here is the patch I described in my previous comment:

```patch
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 979d8e9..da40e41 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -57,7 +57,6 @@ endif()
 # before adding source files via "target_sources"
 # create target binary "idefix"
 add_executable(idefix)
-add_subdirectory(src build)
 
 if(EXISTS ${PROJECT_BINARY_DIR}/setup.cpp)
   target_sources(idefix PUBLIC ${PROJECT_BINARY_DIR}/setup.cpp)
@@ -72,6 +71,7 @@ if(EXISTS ${PROJECT_BINARY_DIR}/CMakeLists.txt)
   add_subdirectory(${PROJECT_BINARY_DIR} build/setup)
 endif()
 
+add_subdirectory(src build)
 
 
 if(Idefix_MHD)
```
Pull request to follow...

**Comment by jeankpf** [comment]
I propose we do not close this issue right now, since I might have found another fix for this bug (see my comment on the unsuccessful pull request linked to this issue).

**Comment by neutrinoceros** [comment]
fine by me

**Comment by glesur** [comment]
In principle, custum CMakeLists.txt are here for setup-specific **physics** module, like MHD, vector potential, EOS, etc... (i.e. you don't forget to switch on MHD for an MHD setup).

Here, we're talking about an **architectural** option (MPI). I don't want to allow this kind of option in setup CMakeLists.txt because it prevents setups from being portable (it won't compile if MPI is not installed). And if your setup is bugged and you want to remove MPI to debug, then it becomes problematic.

If you need options like this, then they should be attached to a particular cluster (because it's architectural, not physical). Best practice is then to store all of the architecture options in an environement variable that is initialised when you connect to said cluster. e.g.:
`export IDEFIX_OPTS=-DIdefix_MPI=ON -DKokkos_ENABLE_CUDA=ON -DKokkos_ARCH_AMPERE80=ON`
And then
`cmake $IDEFIX_DIR $IDEFIX_OPTS`

**Comment by glesur** [comment]
Despite my comment above, this bug is now fixed in #186 so that the behaviour of MPI is consistent with other options.

I however still do not recommend enabling MPI in a user setup CMakeLists.txt.
